consensus/src/quorum_store/tests/quorum_store_wrapper_test.rs

```diff
@@ -0,0 +1,163 @@
+use aptos_types::transaction::SignedTransaction;
+use bcs::to_bytes;
+use consensus_types::common::{Payload, PayloadFilter, TransactionSummary};
+use consensus_types::proof_of_store::{LogicalTime, ProofOfStore, SignedDigestInfo};
+use consensus_types::request_response::{ConsensusResponse, WrapperCommand};
+use futures::channel::mpsc::Receiver;
+use futures::channel::{mpsc, oneshot};
+use futures::StreamExt;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::timeout;
+
+async fn queue_mempool_batch_response(
+    txns: Vec<SignedTransaction>,
+    quorum_store_to_mempool_receiver: &mut Receiver<QuorumStoreRequest>,
+) -> Vec<TransactionSummary> {
+    if let QuorumStoreRequest::GetBatchRequest(_max_batch_size, exclude_txns, callback) = timeout(
+        Duration::from_millis(1_000),
+        quorum_store_to_mempool_receiver.select_next_some(),
+    )
+    .await
+    .unwrap()
+    {
+        callback
+            .send(Ok(QuorumStoreResponse::GetBatchResponse(txns)))
+            .unwrap();
+        exclude_txns
+    } else {
+        panic!("Unexpected variant")
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_batch_creation() {
+    let (quorum_store_to_mempool_tx, mut quorum_store_to_mempool_rx) = mpsc::channel(1_024);
+    let (wrapper_quorum_store_tx, mut wrapper_quorum_store_rx) = tokio::sync::mpsc::channel(100);
+
+    let mut wrapper = QuorumStoreWrapper::new(
+        0,
+        Arc::new(DummyDataManager::new()),
+        quorum_store_to_mempool_tx,
+        wrapper_quorum_store_tx,
+        10_000,
+        10,
+        9 * size_of_signed_transaction() as u64,
+    );
+
+    let signed_txns = create_vec_signed_transactions(500);
+    signed_txns
+        .iter()
+        .for_each(|txn| assert_eq!(168, to_bytes(txn).unwrap().len()));
+
+    let join_handle = tokio::spawn(async move {
+        let mut num_txns = 0;
+
+        let signed_txns = create_vec_signed_transactions(1);
+        queue_mempool_batch_response(
+            vec![signed_txns[0].clone()],
+            &mut quorum_store_to_mempool_rx,
+        )
+        .await;
+        // Expect AppendToBatch for 1 txn
+        let quorum_store_command = wrapper_quorum_store_rx.recv().await.unwrap();
+        if let QuorumStoreCommand::AppendToBatch(data) = quorum_store_command {
+            assert_eq!(data.len(), signed_txns.len());
+            assert_eq!(data, signed_txns);
+        } else {
+            panic!("Unexpected variant")
+        }
+        num_txns += 1;
+
+        let signed_txns = create_vec_signed_transactions(9);
+        // Expect single exclude_txn
+        let exclude_txns =
+            queue_mempool_batch_response(signed_txns.clone(), &mut quorum_store_to_mempool_rx)
+                .await;
+        assert_eq!(exclude_txns.len(), num_txns);
+        // Expect EndBatch for 1 + 9 = 10 txns. The last txn pulled is not included in the batch.
```

Comment:
It's interesting that the expected behavior is now that at least one txn pulled is not included in the batch.

Reply:
This is because mempool doesn't understand bytes, so I think it is unavoidable. But will get this transaction next time we pull, right?